usr/src/uts/common/io/scsi/adapters/smartpqi/smartpqi_init.c

```diff
@@ -0,0 +1,1948 @@
+    pqi_raid_error_info_t e_info);
+static boolean_t identify_controller(pqi_state_t s,
+    bmic_identify_controller_t *ident);
+static boolean_t write_host_wellness(pqi_state_t s, void *buf, size_t len);
+static boolean_t get_device_list(pqi_state_t s, report_phys_lun_extended_t **pl,
+    report_log_lun_extended_t **ll);
+static boolean_t build_raid_path_request(pqi_raid_path_request_t *rqst, int cmd,
+    caddr_t lun, uint32_t len, int vpd_page);
+static boolean_t identify_physical_device(pqi_state_t s, pqi_device_t devp,
+    bmic_identify_physical_device_t *buf);
+static pqi_device_t create_phys_dev(pqi_state_t s,
+    report_phys_lun_extended_entry_t *e);
+static pqi_device_t create_logical_dev(pqi_state_t s,
+    report_log_lun_extended_entry_t *e);
+static boolean_t is_new_dev(pqi_state_t s, pqi_device_t new_dev);
+static boolean_t revert_to_sis(pqi_state_t s);
+static void save_ctrl_mode(pqi_state_t s, int mode);
+static boolean_t scsi_common(pqi_state_t s, pqi_raid_path_request_t *rqst,
+    caddr_t buf, int len);
+static void update_time(void *v);
+
+static int reset_devices = 1;
+
+boolean_t
+pqi_check_firmware(pqi_state_t s)
+{
+	uint32_t	status;
+
+	status = G32(s, sis_firmware_status);
+	if (status & SIS_CTRL_KERNEL_PANIC)
+		return (B_FALSE);
+
+	if (sis_read_scratch(s) == SIS_MODE)
+		return (B_TRUE);
+
+	if (status & SIS_CTRL_KERNEL_UP) {
+		sis_write_scratch(s, SIS_MODE);
+		return (B_TRUE);
+	} else {
+		return (revert_to_sis(s));
+	}
+}
+
+boolean_t
+pqi_prep_full(pqi_state_t s)
+{
+	func_list_t	*f;
+
+	for (f = startup_funcs; f->func_name != NULL; f++)
+		if (f->func(s) == B_FALSE)
+			return (B_FALSE);
+
+	return (B_TRUE);
+}
+
+/*
+ * []----------------------------------------------------------[]
+ * | Startup functions called in sequence to initialize HBA.	|
+ * []----------------------------------------------------------[]
+ */
+
+static boolean_t
+pqi_calculate_io_resources(pqi_state_t s)
+{
+	uint32_t	max_xfer_size;
+	uint32_t	max_sg_entries;
+
+	s->s_max_io_slots = min(PQI_MAX_OUTSTANDING_REQUESTS,
+	    s->s_max_outstanding_requests);
+
+	max_xfer_size = min(s->s_max_xfer_size, PQI_MAX_TRANSFER_SIZE);
+
+	/* ---- add 1 when buf is not page aligned ---- */
+	max_sg_entries = max_xfer_size / PAGESIZE + 1;
+	max_sg_entries = min(max_sg_entries, s->s_max_sg_entries);
+	max_xfer_size = (max_sg_entries - 1) * PAGESIZE;
+
+	s->s_sg_chain_buf_length = (max_sg_entries * sizeof (pqi_sg_entry_t)) +
+	    PQI_EXTRA_SGL_MEMORY;
+
+	s->s_max_sectors = max_xfer_size / 512;
+
+	return (B_TRUE);
+}
+
+static boolean_t
+pqi_check_alloc(pqi_state_t s)
+{
+	if (s->s_max_outstanding_requests > PQI_MAX_OUTSTANDING_REQUESTS)
+		s->s_max_outstanding_requests = PQI_MAX_OUTSTANDING_REQUESTS;
+
+	s->s_error_dma = pqi_alloc_single(s, (s->s_max_outstanding_requests *
+	    PQI_ERROR_BUFFER_ELEMENT_LENGTH) + SIS_BASE_STRUCT_ALIGNMENT);
+	if (s->s_error_dma == NULL)
+		return (B_FALSE);
+
+	return (B_TRUE);
+}
+
+#define	MILLISECOND	1000
```

Comment:
I was surprised not to find these in some common header, maybe if I make a note of it someone else will have better luck finding them.